static/js/tracking.ts

```diff
@@ -0,0 +1,231 @@
+let changesSent = false;
+
+
+export function initializeTracking() {
+    document.addEventListener("DOMContentLoaded", documentLoaded);
+}
+
+
+function documentLoaded() {
+    // attach events
+    document.addEventListener('click', trackEvent);
+    document.addEventListener('change', trackEvent);
+    
+    
+    // initialize variables
+    clickCounts = handleLocalStorage(CLICK_COUNTS);
+    lastActiveTime = handleLocalStorage(LAST_ACTIVE, Date.now());
+    
+    if (isLoggedIn()) {
+        removeTrackingInterval(); // Possibly removing lingering ones.
+        setTrackingInterval(); // Resume with fresh timer
+    }
+}
+
+function removeTrackingInterval() {
+    const storedData = localStorage.getItem(INTERVAL_KEY);
+    if (storedData) {
+        try {
+            const parsedData = JSON.parse(storedData);
+            clearInterval(parsedData.id); // Clear any potentially lingering timer
+            // console.log(parsedData.id, " interval was removed")
+        } catch (error) {
+            console.error("Error parsing tracking interval data:", error);
+        }
+    }
+
+}
+
+function setTrackingInterval() {
+  const timerId = setInterval(checkUserActivity, WAITING_TIME);
+  localStorage.setItem(INTERVAL_KEY, JSON.stringify({ id: timerId, timestamp: Date.now() }));
+}
+
+
+async function trackEvent(event: Event) {
+    // the following check is necessary since some elements issue click and change events.
+    const currentTime = Date.now();
+    const inactiveDuration = currentTime - lastActiveTime;
+    if (inactiveDuration <= 200) {
+        return;
+    }
+    const target = event.target as HTMLElement;
+    // console.log(target, event.type)
+    if (target.matches('button') || target.matches('a') || target.matches('input') || target.matches('select') || target.matches("div")) {
+        let elementIdOrName = target.id;
+
+        if (!elementIdOrName && target.hasAttribute("name")) {
+            elementIdOrName = target.getAttribute("name") || "";
+        }
+
+        if (ELEMENT_TO_TRACK.includes(elementIdOrName)) {
+            clickCounts = handleLocalStorage(CLICK_COUNTS);
+            
+            const page = window.location.pathname;
+
+            const value = (target as HTMLInputElement).value || "";
+
+            clickCounts.push({time: currentTime, id: elementIdOrName, page, extra: value});
+            
+            // console.log(target, clickCounts)
+            // console.log(`Event: ${event.type}, Element ID or Name: ${elementIdOrName}, Click Count: ${clickCounts}`);
+            handleUserActivity(clickCounts);
+        } 
+    }
+}
+
+// Function to handle user activity
+function handleUserActivity(clickCounts: any) {
+    lastActiveTime = handleLocalStorage(LAST_ACTIVE, Date.now());
+    clickCounts = handleLocalStorage(CLICK_COUNTS, clickCounts);
+    changesSent = false;
+}
+
+// Retrieve or set items in local storage.
+function handleLocalStorage(item: string, value: any = undefined) {
```

Comment:
This function is doing too much, and the word "handle" is a bit vague. It seems to be doing triple double duty:

- Loading from localStorage
- Writing to localStorage
- Knowing the types of keys it's reading and writing.

The last bit first: if the value you're storing is a number, `JSON.parse()` and `parseInt()` will do the same thing, so I think you can simplify and always say you `JSON.parse()` whatever is in `localStorage` and always `JSON.stringify()` before storing. 

As for the double duty: two different functions for reading and writing are simple and easy to understand:

```ts
function readLocalStorage(key: string, defaultValue: any): any;
function writeLocalStorage(key: string, value: any);

// Extra neat
function readLocalStorage<A>(key: string, defaultValue: A): A;
```

If you really want a single function both for reading and writing, I think it's more clear to have that behavior depend on if the user passes in a value or not, so it sort of behaves like an overloaded function. If they supply a `value` argument, then they want to write. If they don't supply a `value` argument, then they want to read. The difference now depends on what the user passes in AND what is already in `localStorage`, and it makes it hard to reason about what the function is doing.